### DISCUSSION TITLE
adding tagstack support

### DISCUSSION THREAD

**Discussion by tksoh** [eteran/nedit-ng#212] [Ideas]
Original nedit has a patch for adding vim-like tagstack feature on Sourceforge:

https://sourceforge.net/p/nedit/patches/60/

A version implemented with nedit macro is also available. Perhaps something to consider for adding into nedit-ng.

**Comment by tksoh** [comment]
Here's VIM's doc on the feature for reference:

https://vim.fandom.com/wiki/Browsing_programs_with_tags

**Comment by eteran** [comment]
Thanks!, I'll take a look. There definitely is an argument to be made for creating a healthy "macro ecosystem" for nedit-ng by implementing things that could be macros as macros. Though, we may want to figure out some sort of "macro repository" if we **really** want to go that route.

For this one, I'll have to play with the idea and see what "feels right". I can say that the fact that we already have a tag search feature and this would basically be kinda adding a history to that makes me **lean** built-in, but you make a good enough point that I'm not 100% sure which way I'll go with this one.

Either way, this seems like a cool feature, so one way or another, we'll get it in there :-).

**Reply by tksoh** [reply]
I think the macro 'ecosystem' can use the same 'drop-in' approach as the one we discussed on the language mode. An easy-to-integrate macro system will definitely encourage more  participation to enhance nedit's ability, especially for those who are not into hacking the source. Of course that might also call for an upgrade to the macro language to make the macro files more self-contained, including feature such as defining the menu entries and keyboard accel shortcuts (right now we have to do that through the macro editor dialog)

**Reply by eteran** [reply]
makes sense. regardless of whether this particular feature ends up being a macro or built-in, we'll definitely want to see if there is a clean way to introduce macros/languages/etc as broken-up files in a directory so they can be shared on an individual basis.

**Reply by tksoh** [reply]
If we could improve the macro ecosystem, then maybe we should push out the integration of tagstack, and make the macro version a temp solution for anyone who might be interest. Look at it as a kind of add-on perhaps. 

The main issue for the macro version is that it's unable to override the Ctrl-D shortcut, although it's using Ctrl left/right brackets like vim does.

**Reply by eteran** [reply]
Right, it may make more sense to prioritize sharable macros over this specific feature. Your macro implementation is nice and simple. 

If I understand the code correctly, it basically just leverages the existing bookmarking system, right? If so, that's a clever choice since the work for maintaining the correct offsets is done for you. But, does that mean that it is limited to a depth of 26 (assuming the user doesn't use bookmarks themselves)?

**Reply by tksoh** [reply]
The macro indeed utilizes the bookmarks. Since it's using the number characters '1'-'9' as markers, it's limited to depth of 9 per calling doc (where you call Find Def). Users are only using 'a'-'z' for regular bookmarks, so they should not interfere with each others.

Tagstack are usually short lived, i.e. in most situation users would pop them to return soon after they view the definitions, else they will lose track of the stack fairly soon. 

It may looks like a weakness, but over the years of using the tagstack, I have not had the need to make them more powerful.

**Reply by eteran** [reply]
Ah, very cool, Well for starters, I'll have to double-check that I didn't explicitly limit bookmarks to be letters! if I did, then that would be a limitation introduced where there need not be one! I don't know that I did that, but I can see myself thinking "well they are only supposed to be letters, so why not enforce that at every level"

**Reply by tksoh** [reply]
Perhaps the easiest way is to just run the macro to see if they work as expected :-)

**Reply by tksoh** [reply]
I just tried to run the macro on NG. Some bad news when trying pop the stack:

1. first it complains that raise_window() function is not available
2. after commenting out raise_window(), it's still unable to return to the calling position
3. I always get a beep when pushing (it's able to find the tag) and popping stack

**Reply by eteran** [reply]
OK, well all 3 of things are worth fixing, I'm on the case!

**Reply by eteran** [reply]
I can say that both `raise_window` and `focus-pane` are commented out in my macro tables with "do these make sense to support?" And I suppose the answer is yes, LOL. So that is the reason for the first issue.

I think the main reason why I wasn't sure if theyt made sense is that nedit classic had a pretty blended concept of a document vs a window under the hood, while NG has a pretty clear distinction between the two. And they seemed a little too window centric. 

Additionally, I believe in classic nedit that essentially any action connected to an event is technically callable by a macro, while NG is selective since Qt doesn't have as generic of a "call this action" concept (it gets close with signals and slots though!). So for example, technically, all the various mouse events are callable from a macro... but since they are mouse events, I don't bother making them "macro-able".

That being said, if there's a use for a given action, it shall be implemented!

**Reply by eteran** [reply]
As for the beep... well I did exactly what I thought I might have!

My "add mark" function insists on a letter and will beep on anything else. I can loosen that requirement, which if done right, will effectively reserve non-letter entries for macro use.

**Reply by eteran** [reply]
@tksoh I think problems 2 and 3 **should** be fixed in this PR: https://github.com/eteran/nedit-ng/pull/218

I'm about to call it a night, so I haven't actually tested it, but it looks right to me :-).

Once I have an opportunity tomorrow to confirm the correctness, I'll merge into master, but at the very least, it's something you can play with until then to continue testing things if you like.

**Reply by tksoh** [reply]
If you view NG as a new version of nedit, then obviously all features supported by nedit 5 need to be present or at least backwards compatible, even though they might not be meaningful due to the new toolkit/environment. This is particularly critical for the macro users.

**Reply by tksoh** [reply]
BTW, the tabbed editing feature was added on top of existing nedit architecture (raise_window and focus-pane were already there before that), and it was intentionally made to allow to the macros to run as-is. The document vs windows distinction is a matter of GUI appearance. Macro does not see the difference in most, if not all, situation.

**Reply by tksoh** [reply]
I will try out the new change if I get the opportunity.

**Reply by tksoh** [reply]
I tried your PR. Popping the stack appears to be working, but without raise_window support, it will naturally only work within the same document.

**Reply by eteran** [reply]
@tksoh  OK, bookmark fix tested and now merged :-)

> If you view NG as a new version of nedit, then obviously all features supported by nedit 5 need to be present or at least backwards compatible, even though they might not be meaningful due to the new toolkit/environment. This is particularly critical for the macro users.

I do of course view NG as the new version of nedit :-). So I agree, that NG needs to be as backward compatible as possible.

Though, there is of course subtlety to this. For example, some things that are technically available to macros aren't even documented, so it's hard to tell if they were even intended to be usable, or if they are usable simply due to implementation details happening to make it work.

For example, technically, a macro in nedit-classic can call `toggle_overstrike`... but I'm unsure if this is omited from the docs because someone forgot to update the docs, or because no one even knew it was technically callable :-P.

None the less, the number of macro functions currently left unimplemented is pretty small. The bottom line is that I agree with you, and will implement all of the ones that can sanely be implemented, even if that implementation is simply returning success and printing a debug message saying that usage is deprecated.

**Comment by eteran** [comment]
[tagstack-v1_0.menu.txt](https://github.com/eteran/nedit-ng/files/5721997/tagstack-v1_0.menu.txt)
[tagstack-v1.3-nedit-5.7.patch.txt](https://github.com/eteran/nedit-ng/files/5721998/tagstack-v1.3-nedit-5.7.patch.txt)

Just adding a quick copy of both the patch and macro file here too for quick reference :-)
